src/bidiMapper/utils/log.ts

```diff
@@ -15,16 +15,18 @@
  */
 export function log(type: string): (...message: any[]) => void {
   return (...messages: any[]) => {
-    const elementId = type + '_log';
+    // If run in browser, add debug message to the page.
+    if (typeof window !== 'undefined' && window?.document?.documentElement) {
+      console.log(type, ...messages);
 
-    if (!window.document.getElementById(elementId)) {
-      window.document.documentElement.innerHTML += `<h3>${type}:</h3><pre id='${elementId}'></pre>`;
-      window.document.getElementById(elementId);
-    }
+      const elementId = type + '_log';
 
-    const element = window.document.getElementById(elementId);
+      if (!window.document.getElementById(elementId)) {
+        window.document.documentElement.innerHTML += `<h3>${type}:</h3><pre id='${elementId}'></pre>`;
+      }
 
-    console.log.apply(null, [type].concat(messages));
-    element.innerText += messages.join(', ') + '\n';
+      const element = window.document.getElementById(elementId);
+      element.innerText += messages.join(', ') + '\n';
```

Comment:
This will get slower and slower with more logs as each time everything needs to be parsed again. Perhaps create and append a `<div>`?

Reply:
fixed